File: 11. Sorting/Bubble Sort/bubbleSort.cpp
#include <iostream>
using namespace std;

void printArr(int arr[], int n)
{
    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
}

void bubbleSort(int arr[], int n)
{

    for (int i = 0; i < n - 1; i++)
    {

        bool isSwap = false;

        for (int j = 0; j < n - i - 1; j++)
        {
            if (arr[j] > arr[j + 1])
            { // if(arr[j] < arr[j+1]) for descending order
                swap(arr[j], arr[j + 1]);
                isSwap = true;
            }
        }
        if (!isSwap)
        {
            return;
        }
        printArr(arr, n); 
    }
}

int main()
{
    int arr[5] = {5, 4, 1, 4, 2};
    bubbleSort(arr, 5);
}

File: 11. Sorting/Bubble Sort/bubbleSort.py
def print_arr(arr):
    print(" ".join(map(str, arr)))

def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        is_swap = False
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:  # change to arr[j] < arr[j + 1] for descending order
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                is_swap = True
        if not is_swap:
            return
        print_arr(arr)

if __name__ == "__main__":
    arr = [5, 4, 1, 4, 2]
    bubble_sort(arr)
